core/primitives/src/cpu.rs

```diff
@@ -0,0 +1,97 @@
+/// Try to ensure that the required CPU features are available and enabled for the current CPU.
+///
+/// Note that this is a best-effort kind of thing. If these features are already enabled when
+/// compiling the program, then we may very well end up executing some unsupported instruction
+/// before we even get to executing this code.
+pub fn ensure_cpu_compatibility(force: bool) {
+    if cfg!(near_production) || force {
+        imp::ensure_cpu_compatibility();
+    } else {
+        // Intentionally empty: allow execution without verifying the properties of the CPU.
+    }
+}
+
+mod imp {
+    const NOTE: &str = "confirm hardware requirements at \
+        <https://docs.near.org/docs/develop/node/validator/hardware>.";
+
+    #[cfg(not(target_arch = "x86_64"))]
+    pub fn ensure_cpu_compatibility() {
+        eprintln!("error: processor running a x86_64 instruction set is required.\nnote: {}", NOTE);
+        std::process::exit(1);
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    pub fn ensure_cpu_compatibility() {
+        // SAFETY: this is “just” a CPUID. Some very ancient chips don't support this instruction
+        // and may do anything else with the opcode.
+        //
+        // As an implementation note, we cannot use `is_*_feature_available!` macro from libstd
+        // here, because it will unconditionally return `true` when `cfg(target_feature="...")` is
+        // true.
+        let cpuid01 = unsafe { std::arch::x86_64::__cpuid(0x01) };
+        ensure_cpuid01(cpuid01.ecx, cpuid01.edx);
+        let cpuid0d = unsafe { std::arch::x86_64::__cpuid(0x0d) };
+        ensure_cpuid0d(cpuid0d.eax);
+    }
+
+    fn ensure_cpuid01(ecx: u32, edx: u32) {
```

Comment:
A link to documentation, where those bits are defined would be nice.
This would make it easier to check, whenever all bits are set correctly while doing the review.